File: apps/web/src/lib/sessions/team-live-runs.ts
import { isCodingSessionStale } from "@exp/db-schema/domain"

import type { CodingSession } from "@/db/schema"
import { sessionDisplayState } from "@/lib/coding-session-display"
import { isLiveRun } from "@/lib/past-runs"

// EXP-1075: session lists are TEAM-scoped, so a run left behind in another
// team is invisible until you switch to it. The team picker carries the
// signal: the caller's OWN live runs, grouped by team, so a dot on the
// switcher (and on each foreign team's row) says "something of yours is
// alive over there". Liveness is `useMyLiveRuns`' rule exactly — live by
// status (or EXP-888's sweep end) AND a fresh heartbeat — so the dot and the
// Running section can never disagree about one run. ×4 lockstep.

/** The fields the grouping reads — structural so a caller holding a partial
 *  row (or a test fixture) still passes one. */
export type TeamLiveRunSession = Pick<
  CodingSession,
  `teamId` | `userId` | `status` | `endedBy` | `needsInput` | `updatedAt`
>

/** One team's live-run summary: how many of MY runs are alive there, and
 *  whether any of them is parked on a question (the amber tone). */
export type TeamLiveRuns = { count: number; needsInput: boolean }

/** The caller's live runs grouped by team. Teams with none are absent from
 *  the map, so `map.get(id)` is the whole "does this team have a dot?" test. */
export function liveRunsByTeam(
  sessions: readonly TeamLiveRunSession[],
  { me, now }: { me: string; now: Date }
): Map<string, TeamLiveRuns> {
  const byTeam = new Map<string, TeamLiveRuns>()
  for (const session of sessions) {
    if (session.userId !== me) continue
    if (!isLiveRun(session)) continue
    if (isCodingSessionStale(new Date(session.updatedAt), now)) continue
    const entry = byTeam.get(session.teamId) ?? { count: 0, needsInput: false }
    entry.count += 1
    // `prState` is not read here: the picker's dot only splits parked-on-me
    // from everything else, and an `in_review` run is never `needs_input`.
    if (sessionDisplayState(session, null) === `needs_input`) {
      entry.needsInput = true
    }
    byTeam.set(session.teamId, entry)
  }
  return byTeam
}

/** The switcher's own dot: the ACTIVE team is excluded — its runs are the
 *  sidebar's Running section, one floor below, and repeating them on the
 *  picker would say "go somewhere else" about work already on screen. */
export function otherTeamsLive(
  byTeam: ReadonlyMap<string, TeamLiveRuns>,
  activeTeamId: string | undefined
): { any: boolean; needsInput: boolean } {
  let any = false
  let needsInput = false
  for (const [teamId, entry] of byTeam) {
    if (teamId === activeTeamId) continue
    if (entry.count === 0) continue
    any = true
    if (entry.needsInput) needsInput = true
  }
  return { any, needsInput }
}

File: apps/android/app/src/main/java/com/exponential/app/domain/TeamLiveRuns.kt
package com.exponential.app.domain

import com.exponential.app.data.db.CodingSessionEntity

/**
 * EXP-1075: live runs of the CALLER's, grouped by team — the input behind the
 * board switcher's "another team has your live runs" dot.
 *
 * Every session surface on the phone is already SELECTED-TEAM scoped
 * (AgentsViewModel.agentRows, AppViewModel.agentsRunning), which means a run
 * started in another team goes entirely dark once the switcher moves. The dot
 * is the one pointer back to it: a DOT, never a count — the number belongs to
 * the list you reach by switching, not to the control that switches.
 */
data class TeamLiveRuns(
    val count: Int,
    /** At least one of this team's live runs is parked on a question/plan. */
    val needsInput: Boolean,
)

/** Whether any team OTHER than the selected one has live runs, and its tone. */
data class OtherTeamsLive(
    val any: Boolean,
    val needsInput: Boolean,
)

/**
 * Own + live sessions grouped by `teamId`. Exactly the predicate
 * [AppViewModel.agentsRunning] applies for the selected team — a teammate's
 * run is never the caller's business (EXP-312: it can't be viewed or steered),
 * and ended/heartbeat-stale rows drop out (EXP-153).
 *
 * `needsInput` uses the DISPLAY state, not the raw flag: EXP-679 lets the
 * server set `needs_input` on every live status, and `in_review` masks it.
 */
fun liveRunsByTeam(
    sessions: List<CodingSessionEntity>,
    me: String?,
    now: Long,
): Map<String, TeamLiveRuns> {
    if (me == null) return emptyMap()
    val out = LinkedHashMap<String, TeamLiveRuns>()
    for (session in sessions) {
        if (session.userId != me) continue
        if (!CodingSessionLiveness.isLive(session, now)) continue
        val needsInput =
            codingSessionDisplayState(session, null) == CodingSessionDisplayState.NeedsInput
        val prev = out[session.teamId]
        out[session.teamId] = TeamLiveRuns(
            count = (prev?.count ?: 0) + 1,
            needsInput = (prev?.needsInput ?: false) || needsInput,
        )
    }
    return out
}

/**
 * Collapses [liveRunsByTeam] to the switcher control's single dot. The
 * SELECTED team is excluded — its runs already light the Agents tab, and a
 * second dot for them would read as "there is somewhere else to look".
 */
fun otherTeamsLive(
    byTeam: Map<String, TeamLiveRuns>,
    selectedTeamId: String?,
): OtherTeamsLive {
    var any = false
    var needsInput = false
    for ((teamId, runs) in byTeam) {
        if (teamId == selectedTeamId) continue
        if (runs.count <= 0) continue
        any = true
        if (runs.needsInput) needsInput = true
    }
    return OtherTeamsLive(any = any, needsInput = needsInput)
}
